# Problem 0

A number is a perfect square, or a square number, if it is the square of a positive integer.
For example, 25
 is a square number because 
5x5=25; it is also an odd square.

The first 5 square numbers are: 
1, 4, 9, 16, 25, and the sum of the odd squares is 1+9+25 = 35
.

Among the first 907 thousand square numbers, what is the sum of all the odd squares?

In [ ]:
squares = [n*n for n in range(1, 907_000) if n%2==1]
sum(squares)

124357107166515500

# Problem 1
 [Source](https://projecteuler.net/problem=1)

If we list all the natural numbers below $10$ that are multiples of $3$ or $5$, we get $3, 5, 6$ and $9$. The sum of these multiples is $23$.

Find the sum of all the multiples of $3$ or $5$ below $1000$.

In [1]:
# Problem 1 workspace
total = 0
for i in range(1, 1000):
    if i%3 == 0 or i%5 == 0:
        total += i
total

233168

## Answer: 233168

___

# Problem 2
 [Source](https://projecteuler.net/problem=2)

Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with $1$ and $2$, the first $10$ terms will be:
$$1, 2, 3, 5, 8, 13, 21, 34, 55, 89, \dots$$

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.

In [3]:
# Problem 2 workspace
fibs = [1,2]
while fibs[-1] < 4e6:
    next = fibs[-1] + fibs[-2]
    fibs.append(next)

even_fibs = [x for x in fibs if x%2 == 0]
sum(even_fibs)

4613732

## Answer: 4613732

___

# Problem 3
 [Source](https://projecteuler.net/problem=3)

The prime factors of $13195$ are $5, 7, 13$ and $29$.

What is the largest prime factor of the number $600851475143$?

In [6]:
# Problem 3 workspace

def naive_factorize(n):
    limit = n**(0.5)
    factors = []
    factor = 2
    while factor <= limit:
        if n%factor == 0:
            factors.append(factor)
            n /= factor
        else:
            factor += 1 #will increment by 1s but will only ever be appended if prime
    return factors

factors = naive_factorize(600851475143)
factors

[71, 839, 1471, 6857]

In [7]:
max(factors)

6857

## Answer: 6857

___

# Problem 4
 [Source](https://projecteuler.net/problem=4)

A palindromic number reads the same both ways. The largest palindrome made from the product of two $2$-digit numbers is $9009 = 91 \times 99$.

Find the largest palindrome made from the product of two $3$-digit numbers.

In [4]:
cat = "cac"
cat == cat[::-1]

True

In [5]:
str(25*4)

'100'

In [14]:
for i in range(9, 5, -2):
    print(i)

9
7


In [19]:
# Problem 4 workspace

def is_palindrome(s):
    return s == s[::-1]


def getLargestPal():
    max_palindrome = 0
    best_pair = (0,0)
    for i in range(999, 99, -1):
        for j in range(999, 99, -1):
            candidate = i*j
            if is_palindrome(str(candidate)) and candidate > max_palindrome:
                max_palindrome = candidate
                best_pair = (i,j)
    return max_palindrome, best_pair
                
                
            
getLargestPal()

(906609, (993, 913))

## Answer: 906609

___

# Problem 5
 [Source](https://projecteuler.net/problem=5)

$2520$ is the smallest number that can be divided by each of the numbers from $1$ to $10$ without any remainder.

What is the smallest positive number that is
**evenly divisible
divisible with no remainder**
by all of the numbers from $1$ to $20$?

In [21]:
# Problem 5 workspace

#20 is low enough to write out prime factors like this by hand :)
x = 2*3*2*5*7*2*3*11*13*2*17*19
x

232792560

In [22]:

for i in range(1, 21):
    print(x%i)

0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0
0


## Answer: 232792560

___

# Problem 6
 [Source](https://projecteuler.net/problem=6)

The sum of the squares of the first ten natural numbers is,

$$1^2 + 2^2 + ... + 10^2 = 385.$$

The square of the sum of the first ten natural numbers is,

$$(1 + 2 + ... + 10)^2 = 55^2 = 3025.$$

Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is $3025 - 385 = 2640$.

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [25]:
# Problem 6 workspace
sum_of_squares = sum([n**2 for n in range(1,101)])
square_of_sum = sum([n for n in range(1, 101)])**2
abs(sum_of_squares - square_of_sum)

25164150

## Answer: 

___

# Problem 7
 [Source](https://projecteuler.net/problem=7)

By listing the first six prime numbers: $2, 3, 5, 7, 11$, and $13$, we can see that the $6$th prime is $13$.

What is the $10\,001$st prime number?

In [ ]:
# Problem 7 workspace

#Some sieve code taken from https://www.geeksforgeeks.org/python/python-program-for-sieve-of-eratosthenes/
#Not the most elegant sieve but good enough for 10k primes

import numpy as np
import math

def sieve_to_limit(limit = 1000, growth_rate = 1.2):
    prime_count = 0
    n = 30
    primes = None
    while prime_count < limit:
        primes = sieve(n, prime_=primes)
        prime_count = len(np.nonzero(primes)[0].tolist())
        n = math.floor(n*growth_rate)
    return np.nonzero(primes)[0].tolist()


def sieve(n=30, prime_=None):
    if prime_ is None:
        prime = np.ones(n + 1, dtype=bool)
        prime[0:2] = False
    else:
        prev_len = len(prime_)
        prime = np.concat((prime_, np.ones(n+1-prev_len, dtype=bool)))
    for p in range(2, int(math.sqrt(n)) + 1):
        if prime[p]:
            prime[p*p:n+1:p] = False
    return prime



In [42]:
big_sieve = sieve_to_limit(10_001)
big_sieve[10_000]

104743

## Answer: 104743

___

# Problem 8
 [Source](https://projecteuler.net/problem=8)

The four adjacent digits in the $1000$-digit number that have the greatest product are $9 \times 9 \times 8 \times 9 = 5832$.

73167176531330624919225119674426574742355349194934
  
96983520312774506326239578318016984801869478851843
  
85861560789112949495459501737958331952853208805511
  
12540698747158523863050715693290963295227443043557
  
66896648950445244523161731856403098711121722383113
  
62229893423380308135336276614282806444486645238749
  
30358907296290491560440772390713810515859307960866
  
70172427121883998797908792274921901699720888093776
  
65727333001053367881220235421809751254540594752243
  
52584907711670556013604839586446706324415722155397
  
53697817977846174064955149290862569321978468622482
  
83972241375657056057490261407972968652414535100474
  
82166370484403199890008895243450658541227588666881
  
16427171479924442928230863465674813919123162824586
  
17866458359124566529476545682848912883142607690042
  
24219022671055626321111109370544217506941658960408
  
07198403850962455444362981230987879927244284909188
  
84580156166097919133875499200524063689912560717606
  
05886116467109405077541002256983155200055935729725
  
71636269561882670428252483600823257530420752963450

Find the thirteen adjacent digits in the $1000$-digit number that have the greatest product. What is the value of this product?

In [46]:
# Problem 8 workspace

digit_string = '''
73167176531330624919225119674426574742355349194934
  
96983520312774506326239578318016984801869478851843
  
85861560789112949495459501737958331952853208805511
  
12540698747158523863050715693290963295227443043557
  
66896648950445244523161731856403098711121722383113
  
62229893423380308135336276614282806444486645238749
  
30358907296290491560440772390713810515859307960866
  
70172427121883998797908792274921901699720888093776
  
65727333001053367881220235421809751254540594752243
  
52584907711670556013604839586446706324415722155397
  
53697817977846174064955149290862569321978468622482
  
83972241375657056057490261407972968652414535100474
  
82166370484403199890008895243450658541227588666881
  
16427171479924442928230863465674813919123162824586
  
17866458359124566529476545682848912883142607690042
  
24219022671055626321111109370544217506941658960408
  
07198403850962455444362981230987879927244284909188
  
84580156166097919133875499200524063689912560717606
  
05886116467109405077541002256983155200055935729725
  
71636269561882670428252483600823257530420752963450

'''

#This can be done more efficiently with regexs but its simple enough to ignore for now
digit_string = digit_string.strip().replace(" ", "").replace("\n", "")
len(digit_string)

1000

In [48]:
digits = [int(c) for c in digit_string]

#We can do this cleverly by storing the product of the middle 11 vals of our scanning 13 digits and only adjusting for the ends but its short enough to brute force (and fewer keystrokes)
best_prod = 0
best_idx = 0

window = 13
for i in range(len(digits)-window + 1):
    prod = 1
    for j in range(i, i+window):
        val = digits[j]
        if val == 0:
            prod = 0
            break
        else:
            prod *= val
    if prod > best_prod:
        best_prod = prod
        best_idx = i

print(best_prod)
print(best_idx)

23514624000
197


In [49]:
digits[197:197+window]

[5, 5, 7, 6, 6, 8, 9, 6, 6, 4, 8, 9, 5]

## Answer: 23514624000

___

# Problem 9
 [Source](https://projecteuler.net/problem=9)

A Pythagorean triplet is a set of three natural numbers, $a \lt b \lt c$, for which,
$$a^2 + b^2 = c^2.$$

For example, $3^2 + 4^2 = 9 + 16 = 25 = 5^2$.

There exists exactly one Pythagorean triplet for which $a + b + c = 1000$.
  
Find the product $abc$.

In [ ]:
# Problem 9 workspace

## Answer: 

___

# Problem 10
 [Source](https://projecteuler.net/problem=10)

The sum of the primes below $10$ is $2 + 3 + 5 + 7 = 17$.

Find the sum of all the primes below two million.

In [ ]:
# Problem 10 workspace

## Answer: 

___